### PR TITLE
Make Unreal lambda bindings on the AgonesComponent safe

### PR DESCRIPTION
**Pull request by achynes**
**What type of PR is this?**
/kind bug

**What this PR does / Why we need it**:
This PR fixes a crash in Unreal projects using the plugin, whenever a World is unloaded (loading a new level).
The issue was with lamda bindings that were being called from the engine's TimerManger, but without passing the calling object, which meant the TimerManager did not know that the object was destroyed and tried to invoke the lambda anyway.
The method we can use instead to avoid this is `BindUObject`, which works with any member UFUNCTION and is invalidation-safe.
Additionally, I replaced the other `BindLambda` calls in the file with `BindWeakLambda`, which is also invalidation-safe. These instances were not causing crashes, but if the ownership of the caller changes in the future, they could potentially crash then, and `BindWeakLambda` is a good practice in general.
EDIT: Added missing clearances of TimerHandles when EndPlay is called also. This would also fix the crashes, but also just serves as state clean-up.

**UE4 Build Output:**
```
1>------ Build started: Project: AgonesTestProject, Configuration: Development_Editor x64 ------
1>Using 'git status' to determine working set for adaptive non-unity build (D:\UE4_Vela_4.24.1).
1>Building AgonesTestProjectEditor and ShaderCompileWorker...
1>Using Visual Studio 2019 14.24.28314 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314) and Windows 10.0.18362.0 SDK (C:\Program Files (x86)\Windows Kits\10).
1>[Upgrade]
1>[Upgrade] Using backward-compatible build settings. The latest version of UE4 sets the following values by default, which may require code changes:
1>[Upgrade]     bLegacyPublicIncludePaths = false                 => Omits subfolders from public include paths to reduce compiler command line length. (Previously: true).
1>[Upgrade]     ShadowVariableWarningLevel = WarningLevel.Error   => Treats shadowed variable warnings as errors. (Previously: WarningLevel.Warning).
1>[Upgrade]     PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs   => Set in build.cs files to enables IWYU-style PCH model. See https://docs.unrealengine.com/en-US/Programming/BuildTools/UnrealBuildTool/IWYU/index.html. (Previously: PCHUsageMode.UseSharedPCHs).
1>[Upgrade] Suppress this message by setting 'DefaultBuildSettings = BuildSettingsVersion.V2;' in AgonesTestProjectEditor.Target.cs, and explicitly overriding settings that differ from the new defaults.
1>[Upgrade]
1>Building 12 actions with 12 processes...
1>  [1/12] Module.Voice.cpp
1>  [2/12] UE4Editor-Voice.lib
1>     Creating library D:\UE4_Vela_4.24.1\Engine\Intermediate\Build\Win64\UE4Editor\Development\Voice\UE4Editor-Voice.lib and object D:\UE4_Vela_4.24.1\Engine\Intermediate\Build\Win64\UE4Editor\Development\Voice\UE4Editor-Voice.exp
1>  [3/12] Module.OnlineSubsystemUtils.gen.cpp
1>  [4/12] UE4Editor-Voice.dll
1>     Creating library D:\UE4_Vela_4.24.1\Engine\Intermediate\Build\Win64\UE4Editor\Development\Voice\UE4Editor-Voice.suppressed.lib and object D:\UE4_Vela_4.24.1\Engine\Intermediate\Build\Win64\UE4Editor\Development\Voice\UE4Editor-Voice.suppressed.exp
1>  [5/12] Module.OnlineSubsystemUtils.cpp
1>  [6/12] UE4Editor-OnlineSubsystemUtils.lib
1>     Creating library D:\UE4_Vela_4.24.1\Engine\Plugins\Online\OnlineSubsystemUtils\Intermediate\Build\Win64\UE4Editor\Development\OnlineSubsystemUtils\UE4Editor-OnlineSubsystemUtils.lib and object D:\UE4_Vela_4.24.1\Engine\Plugins\Online\OnlineSubsystemUtils\Intermediate\Build\Win64\UE4Editor\Development\OnlineSubsystemUtils\UE4Editor-OnlineSubsystemUtils.exp
1>  [7/12] UE4Editor-OnlineSubsystemUtils.dll
1>     Creating library D:\UE4_Vela_4.24.1\Engine\Plugins\Online\OnlineSubsystemUtils\Intermediate\Build\Win64\UE4Editor\Development\OnlineSubsystemUtils\UE4Editor-OnlineSubsystemUtils.suppressed.lib and object D:\UE4_Vela_4.24.1\Engine\Plugins\Online\OnlineSubsystemUtils\Intermediate\Build\Win64\UE4Editor\Development\OnlineSubsystemUtils\UE4Editor-OnlineSubsystemUtils.suppressed.exp
1>  [8/12] Module.Agones.gen.cpp
1>  [9/12] Module.Agones.cpp
1>  [10/12] UE4Editor-Agones.lib
1>     Creating library D:\Repos\AgonesTestProject\AgonesTestProject\Plugins\Agones\Intermediate\Build\Win64\UE4Editor\Development\Agones\UE4Editor-Agones.lib and object D:\Repos\AgonesTestProject\AgonesTestProject\Plugins\Agones\Intermediate\Build\Win64\UE4Editor\Development\Agones\UE4Editor-Agones.exp
1>  [11/12] UE4Editor-Agones.dll
1>     Creating library D:\Repos\AgonesTestProject\AgonesTestProject\Plugins\Agones\Intermediate\Build\Win64\UE4Editor\Development\Agones\UE4Editor-Agones.suppressed.lib and object D:\Repos\AgonesTestProject\AgonesTestProject\Plugins\Agones\Intermediate\Build\Win64\UE4Editor\Development\Agones\UE4Editor-Agones.suppressed.exp
1>  [12/12] AgonesTestProjectEditor.target
1>Total time in Parallel executor: 113.88 seconds
1>Total execution time: 116.05 seconds
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

### CODE REVIEW

**Comment by achynes** [comment]
@googlebot I signed it!

**Comment by achynes** [comment]
Missed a couple of files in my initial push, sorry about that. Have pushed them along with clearance of TimerHandles in EndPlay (see edited PR description).

**Comment by achynes** [comment]
Local builds are looking okay, so not sure what to fix for the Cloud Build failure, or if there's any way I can look at that build log?

**Comment by roberthbailey** [comment]
@domgreen

File: sdks/unreal/Agones/Source/Agones/Private/AgonesComponent.cpp [hunk 1]
```diff
@@ -88,8 +95,8 @@ void UAgonesComponent::ConnectSuccess(const FGameServerResponse GameServerRespon
 void UAgonesComponent::Ready(const FReadyDelegate SuccessDelegate, const FAgonesErrorDelegate ErrorDelegate)
 {
 	TSharedRef<IHttpRequest> Request = BuildAgonesRequest("ready");
-	Request->OnProcessRequestComplete().BindLambda(
-		[SuccessDelegate, ErrorDelegate](FHttpRequestPtr HttpRequest, FHttpResponsePtr HttpResponse, const bool bSucceeded) {
+	Request->OnProcessRequestComplete().BindWeakLambda(
+		this, [SuccessDelegate, ErrorDelegate](FHttpRequestPtr HttpRequest, FHttpResponsePtr HttpResponse, const bool bSucceeded) {
```
Comment (domgreen) [inline sdks/unreal/Agones/Source/Agones/Private/AgonesComponent.cpp:99]: nit - can we keep the style consistent:
```
Request->OnProcessRequestComplete().BindWeakLambda(
		this, [SuccessDelegate
```
vs 
```
Request->OnProcessRequestComplete().BindWeakLambda(this, 
```
Will see what my clang formatter says (if anything)

**Review by domgreen — COMMENTED** [review]
Looks good, pretty straight forward ... thanks for bringing the code quality up (I'm pretty new to Unreal) would like to give it a quick run on Windows (probably tomorrow now).

File: sdks/unreal/Agones/Source/Agones/Private/AgonesComponent.cpp (hunk 1 above)
Comment (achynes) [reply sdks/unreal/Agones/Source/Agones/Private/AgonesComponent.cpp:99]: Yeah our project's clang is what decided on this, but happy to conform to your preference :)

Comment (domgreen) [reply sdks/unreal/Agones/Source/Agones/Private/AgonesComponent.cpp:99]: more that both are used ... super small nit

**Comment by achynes** [comment]
> Looks good, pretty straight forward ... thanks for bringing the code quality up (I'm pretty new to Unreal) would like to give it a quick run on Windows (probably tomorrow now).

Great, let me know whenever you've had time and I'll make changes as required. Logging off in EU time here so apologies if I don't see replies immediately.

**Comment by markmandel** [comment]
Looks like SDK Conformance testing flakiness :+1: 
`{"expected":["ready","allocate","setlabel","setannotation","gameserver","health","shutdown","watch","reserve"],"message":"Testing Failed","received":["ready","watch","gameserver","setannotation","setlabel","health","reserve"],"sdkName":"node","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2020-08-28T10:29:27.56712511Z"}`

@achynes the instructions for looking at builds are posted in the comments above. Let me know if you have issues.

**Comment by achynes** [comment]
> Looks like SDK Conformance testing flakiness 👍
> `{"expected":["ready","allocate","setlabel","setannotation","gameserver","health","shutdown","watch","reserve"],"message":"Testing Failed","received":["ready","watch","gameserver","setannotation","setlabel","health","reserve"],"sdkName":"node","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2020-08-28T10:29:27.56712511Z"}`
> 
> @achynes the instructions for looking at builds are posted in the comments above. Let me know if you have issues.

Hi Mark, thanks for the info. Don't know how I missed that second link with the build logs!
The code I changed was exclusively inside the Unreal Plugin, so I'm not sure there's anything that could have incurred the error that occurred, but I'm also not accustomed to seeing Agones code and logs, so let me know if there's anything I may have missed.

**Comment by markmandel** [comment]
@domgreen if you tell me this is good to go, I'm happy to merge :+1:

**Review by domgreen — APPROVED** [review]
Have ran as part of my codebase and all looks fine 👍 

@achynes my linter also wasnt bothered which version of the styling so let's just arrange it to either one of the styles 👍 

@markmandel good to merge, would prefer style nits are consistent but not a blocker 🎉 

@achynes thanks once again, great to have some more eyes on the plugin ❤️

**Comment by achynes** [comment]
> Have ran as part of my codebase and all looks fine 👍
> 
> @achynes my linter also wasnt bothered which version of the styling so let's just arrange it to either one of the styles 👍
> 
> @markmandel good to merge, would prefer style nits are consistent but not a blocker 🎉
> 
> @achynes thanks once again, great to have some more eyes on the plugin ❤️

My pleasure, great to have a plugin to contribute to in the first place. 😁 
I'll fix up that styling now.

**Comment by achynes** [comment]
Opted for `this` on the first line, to isolate the lambda body in reading.

**Comment by domgreen** [comment]
Looks to be another issue with flaky conformance test :/ would be good to get a re-run from someone with perms.

**Pull request merged**